staking/slash/double-sign.go

```diff
@@ -1,52 +1,301 @@
+		}
+
+		halfB := half.TruncateInt()
+		state.AddBalance(reporter, halfB)
+		slashTrack.TotalSnitchReward.Add(slashTrack.TotalSnitchReward, halfB)
+
+		// we might have had enough in undelegations to satisfy the slash debt
+		if paidOff.Equal(half) {
+			continue
+		}
+		// otherwise we need to dig deeper in their pocket
+		leftoverDebt := amt.Sub(half.Sub(paidOff)).TruncateInt()
+		delegator.Amount.Sub(delegator.Amount, leftoverDebt)
+		paidOff.Add(numeric.NewDecFromBigInt(leftoverDebt))
+		slashTrack.TotalSlashed.Add(slashTrack.TotalSlashed, leftoverDebt)
+
+		if paidOff.Equal(half) {
+			fmt.Println("special case when exactly equal", paidOff.String(), half.String())
+		}
+
+		if !paidOff.GTE(half) {
+			fmt.Println(leftoverDebt.String(), "\n Now have ",
+				delegator.Amount, "-had-",
+				amt.String(),
+				paidOff,
+				half,
+				delegations.String(),
+			)
+
+			return errors.Errorf(
+				"paidoff %s is not >= slash debt %s",
+				paidOff.String(),
+				half.String(),
+			)
+		}
+	}
+	fmt.Println("after delegator slash application", slashTrack.String())
 	return nil
 }
+
+// TODO Need to keep a record in off-chain db of all the slashes?
+
+// Apply ..
+func Apply(
+	chain staking.ValidatorSnapshotReader, state *state.DB,
+	slashes []Record, committee []shard.BlsPublicKey,
+) (*Application, error) {
+	log := utils.Logger()
+	rate := Rate(uint32(len(slashes)), uint32(len(committee)))
+	slashDiff := &Application{big.NewInt(0), big.NewInt(0)}
+
+	for _, slash := range slashes {
+		// TODO Probably won't happen but we probably should
+		// be expilict about reading the right epoch validator snapshot,
+		// because it needs to be the epoch of which the double sign
+		// occurred
+		snapshot, err := chain.ReadValidatorSnapshot(
+			slash.Offender,
+		)
+
+		if err != nil {
+			return nil, errors.Errorf(
+				"could not find validator %s",
+				common2.MustAddressToBech32(slash.Offender),
+			)
+		}
+
+		nowAccountState := state.GetStakingInfo(slash.Offender)
+
+		if nowAccountState == nil {
+			return nil, errors.New("cannot be nil code field in account state")
+		}
+
+		const mustBeValidatorsDelegationToSelf = 1
+		// Handle validator's delegation explicitly for sake of clarity
+		validatorOwnDelegation :=
+			snapshot.Delegations[:mustBeValidatorsDelegationToSelf]
+		if err := delegatorSlashApply(
+			validatorOwnDelegation, rate, state,
+			slash.Reporter, slash.Evidence.Epoch, slashDiff,
+		); err != nil {
+			return nil, err
+		}
+
+		// Now can handle external to the validator delegations, third parties
+		// that trusted the validator with some one token, they must also
+		// be slashed so they have skin in the game
+		rest := snapshot.Delegations[mustBeValidatorsDelegationToSelf:]
+		if err := delegatorSlashApply(
+			rest, rate, state, slash.Reporter, slash.Evidence.Epoch, slashDiff,
+		); err != nil {
+			return nil, err
+		}
+		// ...and I want those that I have not seen before
+		sinceSnapshotDelegations :=
```

Comment:
Why slash the new delegators in current epoch? Their stake are not used in the voting power yet. So their tokens are not dirty.

Reply:
But I remember we had a conversation about the snapshot delegations vs current delegations. Can you elaborate again please what that issue was